### DISCUSSION TITLE
Silly Questions

### DISCUSSION THREAD

**Discussion by dummifiedme** [pymupdf/PyMuPDF#890] [Looking for help] [answered]
Hi again,
I worked on a small code on extracting text and screenshots from the PDF using this library, and now thought of making a few changes in the code. Please help me out on these questions (I couldn't explore better because I have an exam due)

1. How can I have my annotations/comments/drawings in a natural reading order? (I am currently writing all the extracts in a text file as I read them via the `page.getText()` function). I found this [snippet on the wiki](https://github.com/pymupdf/PyMuPDF/wiki/How-to-extract-text-in-natural-reading-order-(up2down,-left2right)), I wanted to know, how can I use it?
2. If I have a highlighted text and some word on the text is underlined, I would like to wrap the underlined part in the extracted text in a symbol (say " " ). How can I acheive that?
    What I could think of it matching the two rects and then going on accordingly, but wouldn't the order mess it up? I mean, it would be difficult to traverse though the text file (the one code is writing), and then put in the wraps around the words when there is a match. 
3. Slightly unrelated: how do I open a pdf to a certain page using file URI? 
    To open a pdf, I tried this `file:///C:/Documents/Sample.pdf` and it opens the PDF. But I am not able to open it to a certain page. For certain page, Adobe manual suggests adding a `#page=3` to the URL (not URI in the manual). Though it was for URL, I tried to use it anyways, and it won't even open the PDF.

**Comment by JorjMcKie** [comment, marked as answer]
Hope I understood well:
To keep the sequence of annotations in reading sequence, you have two options:
1. If you compute the rectangles via text searching (`page.search_for()`), sort those rectangles **_before_** you make the annotations. Sorting rectangles means: pick a rectangle corner (e.g. bottom left one), and then sort by (y, x) coordinates.
2. You can also sort the annotations instead:

```python
annots = [a for a in page.annots()]
annots.sort(key=lambda a: (a.rect.bl.y, a.rect.bl.x))
# now loop through the sorted list of annotations
```

> how do I open a pdf to a certain page using file URI?

Won't work. PyMuPDF is **_not a viewer_** like Adobe Acrobat, but a **_programming libray_**. You must first make a document, before you can load a page.
But why don't you (as a **_programmer_**!) simply make a little function for this?

```python
def opener(filename, pno):
    doc = fitz.open(filename)
    page = doc[pno]
    return doc, page

doc, page = opener("somefile.pdf", 3)
```
Or something like that.

**Reply by dummifiedme** [reply]
Thanks for the reply @JorjMcKie. It was nice to hear form you again.

As for sorting the annotations, I can see how it can be implemented. Thanks :) 

As for the PDF part, I was not asking to open a pdf from the code itself but in general from a text. I wanted something like a click able link using file uri. I basically want to embedd a link, into the markdown file I am generating, to the page of the pdf from where the highlight is  extracted from.

**Reply by JorjMcKie** [reply]
> As for the PDF part, I was not asking to open a pdf from the code itself but in general from a text. I wanted something like a click able link using file uri. I basically want to embedd a link, into the markdown file I am generating, to the page of the pdf from where the highlight is extracted from.

Ah, you mean an external link, that points to some page in anpther PDF! Yes, that is possible: prepre a link dictionary an insert it via `page.insert_link(linkdict)`.
Link kind must be fitz.LINK_GOTOR, you need filename, target page number and the rectangle where the link sits on the source page. Something like this: ``linkdict = {'kind': fitz.LINK_GOTOR, 'file': 'target.pdf', 'page': pno, 'to': target_point, from=rect_on_source_page}``
